### DISCUSSION TITLE
Feature Request: Enforce lower-case branch names

### DISCUSSION THREAD

**Discussion by Strepto** [community/community#20518] [Pull Requests]
As you might know git has a thing where branches are Case Sensitive. When the branches are checked out on a case-insensitive file-system such as Windows this may lead to problems.

Read for instance this stackoverflow issue: https://stackoverflow.com/questions/55051729/github-branches-case-sensitivity-issue

Preventing these kinds of branches by forcing a repository to not accept branches with upper-case names would be a nice way for these kinds of issues to not appear. A nice error message when pushing in the same way as when pushing to a protected branch, could make this issue visible and avoidable.

The problem I have is that sometimes someone writes `Feature/MyFeature` and someone else writes `feature/their-feature`. This makes for instance VS code checkout their feature as `Feature/their-feature` (Note the capital F). When commiting back to that branch , and running "sync" in VS code that will create a new branch on github: `Feature/their-feature`. Switching between these branches get troublesome very fast, as seen in the stack overflow issue. 

I would like to fix this with a repository setting in github, as I do not think that git will resolve this issue any time soon.

Using a branch name pattern will not work, as it is case-insensitive.
![image](https://user-images.githubusercontent.com/3185998/178975425-67e018f2-61b7-4406-b461-78d608aa5798.png)

**Comment by dmstrat** [comment]
I hate that this feature isn't getting traction.  This continues to be a problem across multiple git service implementations like BitBucket and GitHub.  Seriously, this needs a long term fix.

**Reply by david-hunsicker-namely** [reply]
Please. reduce toil. reduce confusion. reduce the noise and nonsense.

**Comment by Strepto** [comment]
I found a hack for enforcing lower-case branch names with the Github Ruleset feature. It's not optimal but may help someone save some time until github solves the branch name capitailzation issues.

The magic is to add `(?-i)` to the regex which is used to disable the case-insensitivity in the standard Ruleset branch restrictions regex!


1. Go to your repo, find the Settings, and find the Rulesets menu item.
2. I created a ruleset targeting all branches.
3. Enable the "Restrict branch names" rule
4. Create a restriction
    1. Applies to: `Branch name` 
    2. Requirement: `Must match a given regex pattern`
    3. Matching pattern: `(?-i)^([a-z0-9\/\-])+$(Branch name must be lower case and may contain / or -)*`
        - The magic is to add the `(?-i)` which is used to disable the case-insensitive regex for the following group `( )`!
        - Also note the Optional description `(Branch name must be lower case and may contain / or -)*`. This helps give a message to the user of what is wrong. (Standard error message displayed in git cli is `remote: - Branch name must not match a given regex pattern: (?-i)^([a-z0-9\/\-])+$`, so adding the description into the regex helps indicate whats wrong)
    4. Description: `Branch name must be fully lowercase` (This description is not available in the git terminal, so not sure where its used)
5. Press `Save Changes` button and test this in your repo

Limits: The above regex suggestion is for ascii characters only. Extending with other character sets is possible, but not needed for me.

**Reply by PavithradeviRajanbabu** [reply]
The description is not displayed in the CLI

**Reply by Strepto** [reply]
> The description is not displayed in the CLI

Correct, thats why the description is repeated in the regex check. It will look something like this in the CLI:

> remote: error: GH013: Repository rule violations found for refs/heads/test/Upper-Case-Branch.
> remote: Review all repository rules at https://github.com/example/repo/rules?ref=refs%2Fheads%2Ftest%2FUpper-Case-Branch
> remote: 
> remote: - Branch name must match a given regex pattern: `(?-i)^([a-z0-9\/\-])+$(Branch name must be lower case and may contain / or -)?`
> remote: 
> To https://github.com/example/repo
>  ! [remote rejected]   test/Upper-Case-Branch -> test/Upper-Case-Branch (push declined due to repository rule violations)

Thats why the suggestion I added above includes the `Branch name must be lower case and may contain / or -` in the regex pattern in addition to the description on github.

**Reply by PavithradeviRajanbabu** [reply]
Thanks for your reply. But my CLI returns:
remote: error: GH013: Repository rule violations found for refs/heads/Upper-Case-Branch.
remote: Review all repository rules at https://github.com/example/repo/rules?ref=refs%2Fheads%2Ftest%2FUpper-Case-Branch
remote:
remote: - **Cannot create ref due to create name restrictions.**
remote:
To https://github.com/example/repo
 ! [remote rejected]         HEAD -> Upper-Case-Branch (push declined due to repository rule violations)
error: failed to push some refs to 'https://github.com/example/repo'

In my GitHub Settings, I have

Branch Targeting criteria: 
Exclude by pattern: (?-i)^([a-z0-9\/\-])+$(Branch name must be lower case and may contain / or -)?
Branch Rules: Restrict Creation

**Reply by Strepto** [reply]
This is what I use. Remark: use the "Must match a given Regex pattern".

<img width="651" height="492" alt="image" src="https://github.com/user-attachments/assets/9dffa687-c292-45fa-aaa4-3100139f6f7d" />

**Reply by PavithradeviRajanbabu** [reply]
Found it hard way. Thanks :) 
also (?-i) doesn't help for me to filter out small cases with GitHub. I used regex with 
[^A-Z]* . It helps when we create the branch directly on GitHub. But when we create and push via git to remote, it matches only one character. For example test, tesT, Test are treated same with git. Cannot find a way to only let users create smaller case branch names.

**Reply by apoorva-sriv** [reply]
`^[a-z0-9\/\-]$` was sufficient for me; there was no need for `(?-i)`.

**Comment by Uxorious** [comment]
If the Github side was to make an official fix, I would personally prefer it to not just force everything to lowercase.
Instead, it could check if there are any other overlapping branchnames with different capitalization, and if so, reject it.
